mos-platform/common/lib/private_typeinfo.cpp

```diff
@@ -0,0 +1,945 @@
+                        break;
+                }
+                // Zero out found flags
+                info->found_our_static_ptr = false;
+                info->found_any_static_type = false;
+                p->search_above_dst(info, dst_ptr, current_ptr, path_below, use_strcmp);
+                found_our_static_ptr |= info->found_our_static_ptr;
+                found_any_static_type |= info->found_any_static_type;
+            } while (++p < e);
+        }
+        // Restore flags
+        info->found_our_static_ptr = found_our_static_ptr;
+        info->found_any_static_type = found_any_static_type;
+    }
+}
+
+// This is the same algorithm as __vmi_class_type_info::search_above_dst but
+//   simplified to the case that there is only a single base class.
+void
+__si_class_type_info::search_above_dst(__dynamic_cast_info* info,
+                                       const void* dst_ptr,
+                                       const void* current_ptr,
+                                       path_accessibility path_below,
+                                       bool use_strcmp) const
+{
+    if (is_equal(this, info->static_type, use_strcmp))
+        process_static_type_above_dst(info, dst_ptr, current_ptr, path_below);
+    else
+        __base_type->devirt_search_above_dst(info, dst_ptr, current_ptr, path_below, use_strcmp);
+}
+
+template <class type_info_most_derived>
+void __class_type_info::invoke_search_above_dst(const __class_type_info *self,
+                                                __dynamic_cast_info *info,
+                                                const void *dst_ptr,
+                                                const void *current_ptr,
+                                                path_accessibility path_below,
+                                                bool use_strcmp) 
+{
+  static_cast<const type_info_most_derived *>(self)->search_above_dst(
+      info, dst_ptr, current_ptr, path_below, use_strcmp);
+}
+
+// This is the same algorithm as __vmi_class_type_info::search_above_dst but
+//   simplified to the case that there is no base class.
+void
+__class_type_info::search_above_dst(__dynamic_cast_info* info,
+                                    const void* dst_ptr,
+                                    const void* current_ptr,
+                                    path_accessibility path_below,
+                                    bool use_strcmp) const
+{
+    if (is_equal(this, info->static_type, use_strcmp))
+        process_static_type_above_dst(info, dst_ptr, current_ptr, path_below);
+}
+
+struct search_funcs {
+  void (*search_above_dst)(const __class_type_info *, __dynamic_cast_info *,
+                           const void *, const void *, path_accessibility,
+                           bool);
+  void (*search_below_dst)(const __class_type_info *, __dynamic_cast_info *,
+                           const void *, path_accessibility, bool);
+};
+
+enum class dyn_class_type : unsigned char {
```

Comment:
Add implementation comments describing why this is necessary.

Reply:
Correct; -fno-rtti means that none of the type_info instances are generated for any of the dynamic classes.

I've added the requested comments.